### PR TITLE
Improve GDExtension Tools Integration with Editor Debug Tooling

### LINKED ISSUE #8777: Improve GDExtension Tooling with Editor Debugger Tools
### Describe the project you are working on

We are working on a complete visual scripting subsystem solution tool for Godot based on GDExtension.

### Describe the problem or limitation you are having in your project

The Editor's debug tooling aspects are very restrictive and limited regarding GDExtension tooling.  In C++ modules, the tooling would have immediate access to things like ScriptDebugger, EditorDebuggerNode, and so on to interact with critical aspects of the debug subsystem, such as setting breakpoints, adding breakpoints to the Editor Debugger Panel, etc.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

This change provides several benefits:

**Setting breakpoints**
Plug-in and tool developers will be able to use `EditorDebuggerSession::set_breakpoint` as a way to access the same code path that the `CodeEdit` widget does by interacting with `EditorDebuggerNode` to toggle breakpoints. This allows tooling to set a breakpoint within the script debugger subsystem and register the UI entry in the breakpoints panel.

**Checking Breakpoints**
Adding the `ScriptDebugger` via the `EngineDebugger` singleton, tooling can now check whether a specific line/source pair is a breakpoint and trigger a debugger break. This can also allow tooling to trigger a break within the editor in the same way that GDScript does when a specific error state or condition is raised while running in the editor's debugger.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

**Setting Breakpoints**

```cpp
void ScriptNode::set_breakpoint(bool p_active) {
  // Registers breakpoint details in the plugin for this script/node pair
  // This is what gets returned by EditorPlugin::_get_breakpoints
  ScriptPlugin::get_singleton()->set_breakpoint(_script, this, p_active);
  // Now trigger breakpoint via the UI
  // This allows setting the breakpoints details in the Editor Panel and toggling breakpoints during runtime.
  ScriptDebuggerPlugin::get_singleton()->set_breakpoint(_script, this, p_active);
}

void ScriptPlugin::_get_breakpoints(const Ref<Script>& p_script, ScriptNode* p_node, bool p_active) {
  const String break_point = format("%s:%d", p_script->get_path(), p_node->get_id());
  if (p_active && !_breakpoints.has(break_point)) {
    _breakpoints.push_back(break_point);
  }
  else if (!p_active && _breakpoints.has(break_point)) {
    _breakpoints.remove(break_point);
  }
}

void ScriptDebuggerPlugin::set_breakpoint(const Ref<Script>& p_script, ScriptNode* p_node, bool p_active) {
  // Lookup the current session managed by the custom EditorDebuggerPlugin
  Ref<EditorDebuggerSession> session = _get_current_session();
  if (session.is_valid()) {
    // Session is valid, set the breakpoint
    // This toggles the breakpoint in the Editor UI Debugger Panel and will pass it to the underlying debugger 
    // if a debugger session is currently active, allowing for dynamic breakpoint registration
    session->set_breakpoint(p_script.get_path(), p_node->get_id(), p_active);
  }
}
```

**Checking breakpoints at runtime**

```cpp
Variant ScriptInstance::execute_step() {
  /* execute script step */
  if (EngineDebugger::get_singleton()->is_active()) {
    ScriptDebugger *debugger = EngineDebugger::get_singleton()->get_script_debugger();
    if (current_context->has_error()) {     
      debugger->debug(ScriptLanguage::get_singleton(), false, false);
    }
    else if (debugger->is_breakpoint(current_node_id, _script->get_path()) {
      debugger->debug(ScriptLanguage::get_singleton(), false, true);
    }
  }
}
```

### If this enhancement will not be used often, can it be worked around with a few lines of script?

No, these APIs are currently not available.

### Is there a reason why this should be core and not an add-on in the asset library?

This relates to exposed Editor functionality.

### PR DESCRIPTION
**Pull request by Naros**
This PR aims to address several issues around GDExtension and Debug Tooling within the Editor.

* *Bugsquad edit, closes: https://github.com/godotengine/godot-proposals/issues/8777*

### CODE REVIEW

**Comment by AThousandShips** [comment]
Please open a [proposal](https://github.com/godotengine/godot-proposals) to track this and discuss the specifics of this feature

**Comment by Naros** [comment]
@AThousandShips done - https://github.com/godotengine/godot-proposals/issues/8777
Will take a look at the formatting issues now.

File: doc/classes/ScriptDebugger.xml
Comment (AThousandShips) [inline doc/classes/ScriptDebugger.xml:1]: Please provide descriptions for this new documentation, at least a basic description of all methods and the class itself 🙂

File: doc/classes/EditorDebuggerSession.xml
Comment (AThousandShips) [inline doc/classes/EditorDebuggerSession.xml:1]: Same here

File: doc/classes/ScriptDebugger.xml
Comment (Naros) [reply doc/classes/ScriptDebugger.xml:1]: Thanks for the help and guidance, I think it should be set I hope.

**Comment by Naros** [comment]
Hi, would someone me able to tell me whether the failure is related to my change or something that pre-exists in `main`?

File: core/register_core_types.cpp [hunk 1]
```diff
@@ -290,6 +291,7 @@ void register_core_types() {
 	_classdb = memnew(core_bind::special::ClassDB);
 	_marshalls = memnew(core_bind::Marshalls);
 	_engine_debugger = memnew(core_bind::EngineDebugger);
+	_script_debugger = memnew(core_bind::ScriptDebugger);
```
Comment (AThousandShips) [inline core/register_core_types.cpp:294]: You need to delete this, it doesn't happen automatically, see below

**Comment by dsnopek** [comment]
Thanks!

At a high-level, I think this is a great addition, since it's making something that's possible via Godot modules also be possible via GDExtension (and addons).

I'm not really familiar with the debugger API, but for the most part, the code here looks like a thin layer over `ScriptDebugger` from `core/debugger/script_debugger.h`, which seems good to me. But it should probably be reviewed by someone more familiar with this API (@Faless?) to see if the API that's being exposed really makes sense.

**Comment by Naros** [comment]
I'd also add that if there are other bits from `SccriptDebugger` that make sense to include, let me know.  I basically grabbed just what I needed for the short-term but perhaps there are others that may warrant inclusion.

**Review by vnen — APPROVED** [review]
I'm not the most familiar with the debugger code, but this is mostly exposing what's there for extensions. Overall it looks fine to me.

**Review by AThousandShips — APPROVED** [review]

**Comment by akien-mga** [comment]
Could you squash the commits? See [PR workflow](https://docs.godotengine.org/en/latest/contributing/workflow/pr_workflow.html#the-interactive-rebase) for instructions.

**Review by Faless — COMMENTED** [review]
Do we really want to expose `ScriptDebugger` as a separate class?

I feel that class exists in core mostly for historical reasons.

Does it make sense to move the exposed functions directly to `EngineDebugger`?

E.g.:
```
core_bind::EngineDebugger::insert_breakpoint
```

**Comment by Naros** [comment]
> Do we really want to expose `ScriptDebugger` as a separate class?
> 
> I feel that class exists in core mostly for historical reasons.
> 
> Does it make sense to move the exposed functions directly to `EngineDebugger`?

I have to agree with David, really entirely your call, I can certainly align those there if you'd prefer.

**Comment by Faless** [comment]
> Unfortunately, you may be the person most qualified to answer the questions you pose. :-)

> I have to agree with David, really entirely your call, I can certainly align those there if you'd prefer.

Given there doesn't seem to be any objection, I will say let's do that.

It's already confusing enough to have `EngineDebugger` and `EditorDebugger`, adding a `ScriptDebugger` will make it even worse.

I already had plans to move more of the debugger default features to the plugin interface (while still shipping them as part of Godot), so I will take a chance there to see if we can get rid of the whole ScriptDebugger concept internally too.

**Comment by Naros** [comment]
Sounds good, then I'll amend the PR over the weekend and ping you when its ready for a review from your side on the particulars to make sure that the direction and goals are aligned to what you have in mind 👍 @Faless

**Comment by Naros** [comment]
Hi @Faless, I have a few questions about this work.

I was reviewing and testing my changes in a local build to make sure I hadn't missed any additions needed for this to work well for our script language plug-in, and I came across a hack I had used to integrate with the `goto_line` signal.

```gdscript
TypedArray<Node> nodes = editor_node->find_children("*", "EditorDebuggerNode", true, false);
if (nodes.size() == 1) {
  // Wire this signal so when user selects a frame we can handle that here.
  Node* editor_debugger_node = Object::cast_to<Node>(nodes[0]);
  editor_debugger_node->connect("goto_script_line", callable_mp(this, &OrchestratorScriptView::_goto_script_line));
}

The `ScriptEditor` hooks into this directly into the engine to advance the developer position in the script view to the correct line based on the chosen breakpoint; however, as our plug-in isn't based on the `ScriptEditor` classes, we cannot take advantage of that callback.  I want to find a reasonable way to expose this signal.  

What I was considering was adding a new method on `EditorDebuggerPlugin`, i.e.:
```
void goto_script_line(const Ref<Script>& p_script, int p_line);
```

Then, inside `EditorDebuggerNode::_error_selected` and `_text_editor_stack_goto` where, we emit the `goto_script_line` signal that we iterate the registered debugger plugins and dispatch to this new method, allowing the plug-in an opportunity to react.  There may be other alternatives for this, but it seemed like the most likely candidate.

The last task I have open is to check how GDScript integrates with the Breakpoints bottom panel.  I'd like for my visual script tooling to add the breakpoints to that panel much like how the gutter breakpoints work in GDScript.  I should hopefully have that worked out tomorrow with any last questions before this is ready.

**Comment by Naros** [comment]
The build failures seem to be unrelated as `master` seems to be unstable atm.

**Comment by Naros** [comment]
@akien-mga @Faless is it possible to still get this into 4.3?  It would be super helpful for us to have debugger support in the editor and this is a blocker for that.

**Comment by AThousandShips** [comment]
It's marked for 4.3 so unless that is changed it's still planned for 4.3, wait and see

File: doc/classes/EngineDebugger.xml [hunk 2]
```diff
@@ -9,6 +9,32 @@
...
+		</method>
+		<method name="debug">
+			<return type="void" />
+			<param index="0" name="can_continue" type="bool" default="true" />
+			<param index="1" name="is_error_breakpoint" type="bool" default="false" />
+			<description>
+				Starts a debug break in script execution, optionally specifying whether the program can continue based on [param can_continue] and whether the break was due to a breakpoint.
+			</description>
+		</method>
+		<method name="get_depth" qualifiers="const">
```
Comment (Faless) [inline doc/classes/EngineDebugger.xml:26]:
```suggestion
		<method name="get_depth" qualifiers="const" experimental="">
```

File: doc/classes/EngineDebugger.xml [hunk 3]
```diff
@@ -9,6 +9,32 @@
...
+			<return type="int" />
+			<description>
+				Returns the current debug depth.
+			</description>
+		</method>
+		<method name="get_lines_left" qualifiers="const">
```
Comment (Faless) [inline doc/classes/EngineDebugger.xml:32]:
```suggestion
		<method name="get_lines_left" qualifiers="const" experimental="">
```

File: doc/classes/EngineDebugger.xml [hunk 4]
```diff
@@ -78,6 +149,20 @@
 				Sends a message with given [param message] and [param data] array.
 			</description>
 		</method>
+		<method name="set_depth">
```
Comment (Faless) [inline doc/classes/EngineDebugger.xml:152]:
```suggestion
		<method name="set_depth" experimental="">
```

File: doc/classes/EngineDebugger.xml [hunk 5]
```diff
@@ -78,6 +149,20 @@
...
+			<return type="void" />
+			<param index="0" name="depth" type="int" />
+			<description>
+				Sets the current debugging depth.
+			</description>
+		</method>
+		<method name="set_lines_left">
```
Comment (Faless) [inline doc/classes/EngineDebugger.xml:159]:
```suggestion
		<method name="set_lines_left" experimental="">
```

**Review by Faless — APPROVED** [review]
Fantastic job! :trophy: , and sorry for the long delay between reviews.

I was worried of the potential misuse of `set_lines`/`set_depth` which can potentially break debugging but I couldn't figure out a better solution, and it's better to have this than not in 4.3.

So I've approved this, but also suggested marking those methods as experimental.

Thank you for your patience and the great work.

**Comment by Faless** [comment]
> Is it safe then for my `EditorDebuggerPlugin` to cache the `EditorDebuggerSession` or its ID (as I think there should only be one active),

The ID is a bit of an implementation detail, but it's indeed the "tab number".
In general, it should be fine caching between then signals `EditorDebuggerSession.started` and `EditorDebuggerSession.stopped`.

> I also need a way to react to `breakpoint_set_in_tree` and `breakpoints_cleared_in_tree`, which are signals that are also emitted by the `EditorDebuggerNode`, when the user uses the context-menu on the breakpoints in the panel. Again, delegate these to the Plugin or should we consider a different way?

I think indeed those signals could be exposed via the `EditorDebuggerPlugin` given they are "global" (i.e. are relevant to all sessions).

Maybe `goto_script_line` could also be exposed this way?

**Comment by Naros** [comment]
Thanks for the feedback @Faless, I've addressed your recommendations with the documentation and I've added the 3 methods as virtual methods on the `EditorDebuggerPlugin` interface.  Assuming the build succeeds, let me know if there is anything else you would like to see for this.

**Comment by akien-mga** [comment]
Thanks!

**Comment by Faless** [comment]
> I've added the 3 methods as virtual methods on the `EditorDebuggerPlugin` interface.

Oh no... as I mentioned in my comment, those should have been signals not methods :(

**Comment by Naros** [comment]
> > I've added the 3 methods as virtual methods on the `EditorDebuggerPlugin` interface.
> 
> Oh no... as I mentioned in my comment, those should probably have been signals not virtual methods :(

No worries, I can send another PR to follow-up on this if that's acceptable.

**Comment by Faless** [comment]
@Naros I'll let you know in case, it's not the end of the world indeed, just a bit weird, we usually prefer virtual methods for things that needs overriding and either are frequently called (they're faster than signals), or when we need to receive a return (which is not the case for those function).

At the same time, the plugin aims to be "self-contained", so the virtual method can make sense, and might not be worth changing.

Sorry for the noise, I think in the end we'll leave it as it is now :sweat_smile:

**Pull request merged**